Editorial and formatting changes

giocorti opened the pull request:
This PR makes the following editorial and formatting changes to the GTFS specification. 

* Standardized the choice of certain words
  * passenger/customer --> rider
   * Except for “customer service”
  * Blank --> empty
  * Feed vs dataset
  * Field vs field value
  * Transit Organization --> transit agency
   * Note: organization is still used for the group that publishes the feed
  * Itinerary --> Journey (this matches Fares proposal)
* Made editorial changes to enhance clarity and legibility including removing unnecessary and repetitive language. Major rewrites of:
  * `parent_station`
  * `stop_timezone`
  * `arrival_time` and `departure_time`
  * `Monday`,...,`Sunday`
  * `timepoint`
  * `exact_times`
* Formatting changes
  * Added code ticks to denote fields and other code values
  * Standardized Enum option formatting
  * Added hyperlinks to all .txt files and websites
  * Put field types in alphabetical order.
  * Different type setting for examples
* Defined or redefined the following term:
  * Dataset
  * Field Value
  * Record
  * ID
  * Added referencing ID

As you can see, I made a lot of changes and I'm sure we'll need to discuss some of them. Let me know what you think.

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+|  route_id | ID referencing `routes.route_id` | **Required** | Identifies a route. |
+|  service_id | ID referencing `calendar.service_id` or `calendar_dates.service_id` | **Required** | Identifies a set of dates when service is available for one or more routes. |
+|  trip_id | ID | **Required** | Identifies a trip. |
+|  trip_headsign | Text | Optional | Signage text that identifies the trip's destination to riders. Use this field to distinguish between different patterns of service on the same route. If the headsign changes during a trip, `trip_headsign` can be overridden by specifying values for the `stop_times.stop_headsign`. |
```
Subtle change:
"Text that appears on signage identifying the trip's destination to riders."

(I don't know if "signage text" would be universally clear.)

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
-## Feed Files
+- **Color** - A color encoded as a six-digit hexadecimal number. Refer to [https://htmlcolorcodes.com](https://htmlcolorcodes.com) to generate a valid value (the leading "#" is not included). <br> *Ex. `FFFFFF` for white, `000000` for black or `0039A6` for the A,C,E lines in NYMTA.*
+- **Currency Code** - An ISO 4217 alphabetical currency code. For the list of current currency, refer to [wikipedia](https://en.wikipedia.org/wiki/ISO_4217#Active_codes). <br> *Ex. `CAD` for Canadian dollars, `EUR` for euros or `JPY` for Japanese yen.*
+- **Date** - Service day in the YYYYMMDD format. Since time within a service day can be above 24:00:00, a service day often contains information for the subsequent day(s). <br> *Ex. `20180913` for September 13th, 2018.*
```
I suggest we should introduce a separate definition: "Service day"?
Then, the "Date" definition could be referenced in the "Service day" definition.

giocorti replied:
Are you suggesting that we add "Service day" as a field type? Or should it be added as a term definition?

antrim replied:
Yes. Term definition.

giocorti replied:
My draft definition of "Service day" is below. Thoughts?

>Service day - A span of time during which service on a route operates continuously. A service day may exceed 24:00:00 if service begins on one day and ends on a the following day. For example, service that runs from 08:00:00 on Friday to 01:00:00 am on Saturday, would be denoted as running from 08:00:00 to 25:00:00 on the Friday service day.

giocorti replied:
Avoid making a hard definition, let agencies decided. Specify time roll-over property. Main goal: warn reader of distinction between service and calendar day.

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
-|   |  | | * **1** - Times are considered exact. |  
-|   |  | | For stop-time entries without specified arrival and departure times, feed consumers must interpolate arrival and departure times. Feed producers may optionally indicate that such an entry is not a timepoint (value=0) but it is an error to mark a entry as a timepoint (value=1) without specifying arrival and departure times. |  
+|  trip_id | ID referencing `trips.trip_id` | **Required** | Identifies a trip.  | 
+|  arrival_time | Time | **Conditionally required** | Arrival time at a specific stop for a specific trip on a route. If there are not separate times for arrival and departure at a stop, enter the same value for `arrival_time` and `departure_time`.<br/><br/>Scheduled stops where the vehicle strictly adheres to the specified arrival and departure times are timepoints. If this stop is not a timepoi ...
```
> For times occurring after midnight on the service day, enter the time as a value greater than 24:00:00 in HH:MM:SS local time for the day on which the trip schedule begins.

This has been moved to the definition for the "Time" field. However, I think we need to bring this statement back into this field definition. Otherwise it is too easy to overlook.

giocorti commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+|  [fare_attributes.txt](#fare_attributestxt)  | Optional | Fare information for a transit agency's routes. |
+|  [fare_rules.txt](#fare_rulestxt)  | Optional | Fare rules for a transit agency's routes. |
+|  [shapes.txt](#shapestxt)  | Optional | Rules for mapping vehicle travel paths. |
 |  [frequencies.txt](#frequenciestxt)  | Optional | Headway (time between trips) for routes with variable frequency of service. |
```
I'm interpreting "variable" to mean that the route starts off with one headway value and at some point changes to another headway value to accommodate demand. In other words buses run every 10 minutes during rush hour and every 20 minutes during other times. In that sense I think "variable" makes sense but perhaps its potentially confusing.

antrim replied:
It's extra confusing because in some cases headway may not be variable. A route may run every 15 minutes for the whole time it is in service.

Alternative, could use some more word-smithing, but shows the idea: "Headway (time between trips) for routes with consistent headways over operating span(s)."

barbeau replied:
Ick.  Simply describing this file is currently impossible with the differing `exact_times` definitions.

How about:

>Headway (time between trips) for both frequency-based service and a compressed representation of scheduled service

antrim replied:
How about:
"Headway (time between trips) for headway-based service or a compress representation of fixed schedule service."

This avoids the "frequency-based" term, which, though it might be useful here because of the GTFS file name, is much much less common in the industry than "headway-based".

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+|  [calendar.txt](#calendartxt)  | **Conditionally required** | Service dates specified using a weekly schedule with start and end dates. This file is required unless all dates of service are defined in [calendar_dates.txt](#calendar_datestxt). |
+|  [calendar_dates.txt](#calendar_datestxt)  | **Conditionally required** | Exceptions for the services defined in the [calendar.txt](#calendartxt). If [calendar.txt](#calendartxt) is omitted, then [calendar_dates.txt](#calendar_datestxt) is required and must contain all dates of service. |
+|  [fare_attributes.txt](#fare_attributestxt)  | Optional | Fare information for a transit agency's routes. |
+|  [fare_rules.txt](#fare_rulestxt)  | Optional | Fare rules for a transit agency's routes. |
```
Actually, I notice that the proposed modifications later refer to "journeys".

Therefore, for consistency, the change could be "Rules to apply fares for rider journeys."

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
-|   |  | | **Note:** You may list exceptions for particular dates, such as holidays, in the [calendar_dates.txt](#calendar_datestxt) file. |
-|  start_date | Date | **Required** | The **start_date** field contains the start date for the service.  The start_date field's value should be in YYYYMMDD format. |
-|  end_date | Date | **Required** | The **end_date** field contains the end date for the service. This date is included in the service interval.  The **end_date** field's value should be in YYYYMMDD format. |
+|  service_id | ID | **Required** | Identifies a set of dates when service is available for one or more routes. |
```
When there are primary keys in other tables, we often declare that with the term "uniquely identifies". Let's do that here. One reason this is important is that the original text _"Each service_id value can appear at most once in a calendar.txt file."_ **may** have been introduced to clear up an earlier ambiguity about whether a service_id could be shared across entries in calendar.txt. (I made this mistake early on, and I still wish the GTFS allowed the service_id to be shared, but alas that is not allowed…)

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
-|  start_date | Date | **Required** | The **start_date** field contains the start date for the service.  The start_date field's value should be in YYYYMMDD format. |
-|  end_date | Date | **Required** | The **end_date** field contains the end date for the service. This date is included in the service interval.  The **end_date** field's value should be in YYYYMMDD format. |
+|  service_id | ID | **Required** | Identifies a set of dates when service is available for one or more routes. |
+|  monday | Enum | **Required** | Indicates whether the service has been added or removed for all Mondays in the date range specified by the `start_date` and `end_date` fields. Note that exceptions for particular may be listed in [calendar_dates.txt](#calendar_datestxt). Valid values are: |
```
Change: "Indicates whether the service operates on all Mondays in the date range specified by the `start_date` and `end_date` fields."

[Justification: I'm concerned about the use of the words "added" and "removed" here, because they have a connotation of an exception to the base schedule, which is what calendar_dates.txt is for.]

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
-|   |  | | * **(empty)** - If this field is empty, unlimited transfers are permitted. |
-|  agency_id | ID | Optional | Required for feeds with multiple agencies defined in the agency.txt file. Each fare attribute must specify an agency_id value to indicate which agency the fare applies to. |
-|  transfer_duration | Non-negative integer | Optional | The **transfer_duration** field specifies the length of time in seconds before a transfer expires.  When used with a **transfers** value of 0, the **transfer_duration** field indicates how long a ticket is valid for a fare where no transfers are allowed. Unless you intend to use this field to indicate ticket validity, **transfer_duration**  ...
+|  fare_id | ID | **Required** | Identifies a fare class. |
```
Thinking aloud: "Fare class" might be a somewhat misleading term, but I'm thinking we will not address it right now because we are anticipating doing future work on fares.

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
+|  destination_id | ID referencing `stops.zone_id` | Optional | Identifies a destination zone. If a fare class has multiple destination zones, create a record in [fare_rules.txt](#fare_rules.txt) for each `destination_id`. |
+|   |  | | *Ex. The `origin_id` and `destination_id` fields could be used together to specify that fare class "b" is valid for travel between zones 3 and 4, and for travel between zones 3 and 5, the [fare_rules.txt](#fare_rules.txt) file would contain these records for the fare class:* <br>`fare_id,...,origin_id,destination_id` <br>`b,...,3,4`<br> `b,...,3,5` |
+|  contains_id | ID referencing `stops.zone_id` | Optional | Identifies the zones that a rider will enter while using a given fare class. Used in some systems to calculate correct fare class. |
+|   |  | | *Ex. If fare class "c" is associated with all travel on the GRT route that passes through zones 5, 6, and 7 the [fare_rules.txt](#fare_rules.txt) would contain these records:* <br> `zone_id,route_id,...,contains_id` <br>  `c,GRT,...,5` <br>`c,GRT,...,6` <br>`c,GRT,...,7` <br> *Because all `contains_id` zones must be matched for the fare to apply, an journey that passes through zones 5  ...
```
@giocorti I propose that we bring https://code.google.com/archive/p/googletransitdatafeed/wikis/FareExamples.wiki directly into this GitHub repo. The markdown on the archived Google Code page is not displaying correctly.

giocorti replied:
@antrim I'm assuming the google fare examples linked above should be its own markdown file in the transit/gtfs/spec/en/examples/ folder. Does that sound right?

antrim replied:
@giocorti Yes.

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
 File: **Optional**
 
-This table is intended to represent schedules that don't have a fixed list of stop times. When trips are defined in frequencies.txt, the trip planner ignores the absolute values of the **arrival_time** and **departure_time** fields for those trips in [stop_times.txt](#stop_timestxt). Instead, the **stop_times** table defines the sequence of stops and the time difference between each stop.
+[Frequencies.txt](#frequenciestxt) represents scheduled trips that do not have a fixed list of stop times. Arrival and departure times based from information defined in [frequencies.txt](#frequenciestxt), override the the values of associated `stop_times.arrival_time` and `stop_times.departure_time`.
```
change to:

"[Frequencies.txt](#frequenciestxt) represents trips that operate on regular headways."

[see discussion about the frequencies.txt file above; justification for this change: if exact_times=0, then trips are NOT "scheduled". if exact_times=1, then there is a fixed list of stop times]

barbeau replied:
Per my earlier comment, I'd suggest change to 

>Frequencies.txt represents trips that operate on regular headways (time between trips) for both frequency-based service (exact_times=0) and a compressed representation of scheduled service (exact_times=1).

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
-|   |  | | The value of **exact_times** must be the same for all frequencies.txt rows with the same **trip_id**. If **exact_times** is 1 and a frequencies.txt row has a **start_time** equal to **end_time**, no trip must be scheduled. When **exact_times** is 1, care must be taken to choose an **end_time** value that is greater than the last desired trip start time but less than the last desired tr ...
+|  trip_id | ID referencing `trips.trip_id` | **Required** | Identifies a trip to which the specified frequency of service applies. |
+|  start_time | Time | **Required** | Time at which the first vehicle departs from the first stop of the trip with the specified frequency. |
+|  end_time | Time | **Required** | Time at which service changes to a different frequency (or ceases) at the first stop in the trip. |
```
Change: "Time at which service changes to a different headway (or ceases) at the first stop in the trip."

Justification: "frequency" is indeed a term used for the file name, but the spec refers to headways, which is a more familiar transit term.

antrim commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
 File: **Optional**
 
-Trip planners normally calculate transfer points based on the relative proximity of stops in each route. For potentially ambiguous stop pairs, or transfers where you want to specify a particular choice, use transfers.txt to define additional rules for making connections between routes.
+GTFS consumers calculate transfer points based on allowable time and stop proximity. [Transfers.txt](#transferstxt) specifies additional rules and overrides for selected transfers. 
```
Proposed change: "GTFS-consuming applications identify available transfers in passenger journeys based on allowable time and stop proximity."

What do you think of that? I'm not sure but I think it might be useful to introduce the notion of a passenger journey.

giocorti replied:
I'm not sure that we need to specify that transfers occur during rider journeys. Is there anywhere else a transfer could occur?

giocorti commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+* **Record** - A basic data structure comprised of a number of different fields describing a single entity (e.g. transit agency, stop, route, etc.). Represented, in a table, as a row. 
+* **Field Value** - An individual entry in a field. Represented, in a table, as a single cell. 
+* **Required** - The field must be included in the dataset, and a value must be provided for each field. Some required fields permit an empty string as a value (denoted in this specification as "empty"). To enter an empty string, just omit any text between the commas for that field. Note that `0` is interpreted as "a string of value 0", and is not an empty string. See the field definition for det ...
+* **Optional** - The field may be omitted from the dataset. If an optional column is included, each field in the dataset must have a value for that column. An empty string may be included as a value for fields that do not have values for the column. Some optional fields permit an empty string as a value. To enter an empty string, just omit any text between the commas for that field. Note that `0` ...
```
If we're eliminating the phrase 
>"Note that `0` is interpreted as "a string of value 0", and is not an empty string."
in the *Required* definition, shouldn't we also eliminate it here?

LeoFrachet replied:
> each field in the dataset must have a value for that column.

Wait. What. I don't get it. No, it's optional.

LeoFrachet replied:
We should not consider empty as being a value. empty is a lack of value. Otherwise sentences are bitting their own tails. 

No value = empty value.
Required value = non-empty value.

barbeau replied:
>each field in the dataset must have a value for that column.

I'm not sure about their either.  I agree that the value should be optional (not required), but IIRC there are some places in the GTFS spec where this doesn't currently hold true (i.e., optional fields, if provided, must have values).

@giocorti If the above is true, could you list some examples?

giocorti replied:
@barbeau In the current version of the spec `arrival_time` and `departure_time` both say
>If interpolated times are indicated with timepoint=0, then time points must be indicated with a value of 1 for the timepoint field.

I'll let you know if I find more examples.

giocorti replied:
> We should not consider empty as being a value. empty is a lack of value. Otherwise sentences are bitting their own tails.
> 
> No value = empty value.
> Required value = non-empty value.

If we do this I think it'll create a few issues that we'll need to settle:
1. We'll need clear language to differentiate between a field value having no value because it is empty and a field value having no value because the field is omitted. 
2. On a related note we'll need to address whether GTFS-consuming applications should ever make different interpretations about omitted vs. empty field values. 
3. We'll need to rewrite a few of the definitions such as the "required" definition. I'm currently working on this and I'll float a few ideas soon. 

@LeoFrachet, @barbeau thoughts? Especially on the first two?

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
-
-## Feed Files
+- **Color** - A color encoded as a six-digit hexadecimal number. Refer to [https://htmlcolorcodes.com](https://htmlcolorcodes.com) to generate a valid value (the leading "#" is not included). <br> *Ex. `FFFFFF` for white, `000000` for black or `0039A6` for the A,C,E lines in NYMTA.*
+- **Currency Code** - An ISO 4217 alphabetical currency code. For the list of current currency, refer to [wikipedia](https://en.wikipedia.org/wiki/ISO_4217#Active_codes). <br> *Ex. `CAD` for Canadian dollars, `EUR` for euros or `JPY` for Japanese yen.*
```
I'm reluctant to hide data (like the reference list of the ISO 4117) under a URL. I'd rather have what you've done above with the HTML color code, and keep the linked displayed.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
-* **Dataset unique** - The field contains a value that maps to a single distinct entity within the column. For example, if a route is assigned the ID **1A**, then no other route may use that route ID. However, you may assign the ID **1A** to a location because locations are a different type of entity than routes.
+* **Feed** - Successive datasets comprise a feed.
+* **Dataset** - A complete set of files defined by this specification reference. Datasets should be published at a public, permanent URL, including the zip file name. (e.g., www.agency.org/gtfs/gtfs.zip). For more information see GTFS best practices. 
+* **Field** - A specific type of information for records in a .txt file. Represented, in a table, as a column.
```
> A specific type of information

I don't get it.

I think I would define record, then field, then field value.

barbeau replied:
IMHO a "field" is a property of some object or entity, it's not a type of information.

barbeau replied:
Agreed that Field and Field Value definitions should be together.  Personally, I'd put Record after these two, but that's debatable.

giocorti replied:
> > A specific type of information
> 
> I don't get it.

You're right, the wording needs to changed. "Specific type of information" is supposed to get at the fact that all the field values of a particular field describe a specific property of some object . In other words, the field `stop_lat` always contains a specific type of geospatial information about a location. Perhaps this doesn't need to be included, but then I'm not sure what we should say about fields other than the fact that they are often represented as columns. 

Any ideas on how we should define "field"?

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+* **Field** - A specific type of information for records in a .txt file. Represented, in a table, as a column.
+* **Record** - A basic data structure comprised of a number of different fields describing a single entity (e.g. transit agency, stop, route, etc.). Represented, in a table, as a row. 
+* **Field Value** - An individual entry in a field. Represented, in a table, as a single cell. 
+* **Required** - The field must be included in the dataset, and a value must be provided for each field. Some required fields permit an empty string as a value (denoted in this specification as "empty"). To enter an empty string, just omit any text between the commas for that field. Note that `0` is interpreted as "a string of value 0", and is not an empty string. See the field definition for det ...
```
"for each field" ? You mean for each "field value", no?

> Some required fields permit an empty string as a value

I don't think so. I would call them optional. But maybe I'm missing something

giocorti replied:
>"for each field" ? You mean for each "field value", no?

It should be field value, good catch.

> > Some required fields permit an empty string as a value
> 
> I don't think so. I would call them optional. But maybe I'm missing something

`transfers` in fare_attributes.txt and `transfer_type` in transfers.txt are both listed as required and permit empty field values as an enum option.

LeoFrachet replied:
![giphy](https://user-images.githubusercontent.com/3440724/48020539-244dd700-e104-11e8-8634-84c655e987a6.gif)

Oh no. Why. 

We should change that. It make no sense IMHO. @barbeau @antrim Thoughts? We should discuss that on Thursday.

barbeau replied:
IMHO `and a value must be provided for each field` should be `and a value must be provided in that field for each record`.

antrim replied:
Good catch. We could test fixing that but through a different pull request / discussion.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
-|  [agency.txt](#agencytxt) | **Required** | One or more transit agencies that provide the data in this feed. |
-|  [stops.txt](#stopstxt) | **Required** | Individual locations where vehicles pick up or drop off passengers. |
+|  [agency.txt](#agencytxt) | **Required** | Transit agencies that provide the data in this dataset. |
+|  [stops.txt](#stopstxt) | **Required** | Locations where vehicles pick up or drop off riders. |
```
We should add something about stations and entrances.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
 |  [frequencies.txt](#frequenciestxt)  | Optional | Headway (time between trips) for routes with variable frequency of service. |
 |  [transfers.txt](#transferstxt)  | Optional | Rules for making connections at transfer points between routes. |
-|  [feed_info.txt](#feed_infotxt)  | Optional | Additional information about the feed itself, including publisher, version, and expiration information. |
+|  [feed_info.txt](#feed_infotxt)  | Optional | Feed metadata, including publisher, version, and expiration information. |
```
Should be `Dataset metadata`. The start and end date do not apply to the feed but to the dataset. I know it's confusing...

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
-|  agency_phone | Phone number | Optional | The **agency_phone field** contains a single voice telephone number for the specified agency. This field is a string value that presents the telephone number as typical for the agency's service area. It can and should contain punctuation marks to group the digits of the number. Dialable text (for example, TriMet's "503-238-RIDE") is permitted, but the f ...
-|  agency_fare_url | URL | Optional | The **agency_fare_url** specifies the URL of a web page that allows a rider to purchase tickets or other fare instruments for that agency online. |
-|  agency_email | Email | Optional | Contains a single valid email address actively monitored by the agency’s customer service department. This email address will be considered a direct contact point where transit riders can reach a customer service representative at the agency. |
+|  agency_id | ID | **Conditionally Required** | Identifies a transit agency. A dataset may contain data from multiple agencies. This field is required when the dataset contains data for multiple transit agencies, otherwise it is optional. |
```
If you wanna be by-the-book consistent, all your "descriptions" you either start by a name, or start by a verb. Here your first one is a verb ("identifies") and the second one is a name ("full name").

Since most of them are already names, you can change agency_id to something like "Identifier which..." or "String that..."

You can also leave it as it. But it's less consistent.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
-|   |  | | * **1** - the station entrance is wheelchair accessible (e.g. an elevator is available to platforms if they are not at-grade)  |  
-|   |  | | * **2** - there exists no accessible path from the entrance to station platforms |  
+|  stop_id | ID | **Required** | Identifies a stop, station, or station entrance/exit. Stops, stations or station entrances/exits are collectively referred to as locations. Multiple routes may use the same stop. | 
+|  stop_code | Text | Optional | Short text or a number that identifies the location for riders. These codes are often used in phone-based transit information systems or printed on signage to make it easier for riders to get information for a particular location. The `stop_code` can be the same as `stop_id` if it is public facing. This field should be left "empty" for locations without a code pre ...
```
the double quotes around `empty` are a bit confusing IMHO. Maybe just remove them.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
-|   |  | | * **2** - there exists no accessible path from the entrance to station platforms |  
+|  stop_id | ID | **Required** | Identifies a stop, station, or station entrance/exit. Stops, stations or station entrances/exits are collectively referred to as locations. Multiple routes may use the same stop. | 
+|  stop_code | Text | Optional | Short text or a number that identifies the location for riders. These codes are often used in phone-based transit information systems or printed on signage to make it easier for riders to get information for a particular location. The `stop_code` can be the same as `stop_id` if it is public facing. This field should be left "empty" for locations without a code pre ...
+|  stop_name | Text | **Required** | Name of a location. Use a name that people will understand in the local and tourist vernacular. | 
```
Above you said:

> ... identifies *the* location

And here you say

> ... of *a* location

We likely should always use the same article.

IMHO "the" make sense, since they are properties of an object.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
-|   |  | | * **0** (or empty) - the station entrance will inherit its **wheelchair_boarding** value from the parent station, if specified in the parent |  
-|   |  | | * **1** - the station entrance is wheelchair accessible (e.g. an elevator is available to platforms if they are not at-grade)  |  
-|   |  | | * **2** - there exists no accessible path from the entrance to station platforms |  
+|  stop_id | ID | **Required** | Identifies a stop, station, or station entrance/exit. Stops, stations or station entrances/exits are collectively referred to as locations. Multiple routes may use the same stop. | 
```
I would suggest to add:

> "Station entrances/exits" are commonly refered as just "entrances" in this specification.

Just to lighten the rest of the spec. But up to you.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+|  stop_lon | Longitude | **Required** | Longitude of a location. |  
+|  zone_id | ID | **Conditionally Required** | Identifies the fare zone for a stop. This field is required if providing fare information using [fare_rules.txt](#fare_rulestxt), otherwise it is optional. If this `stop_id` represents a station or station entrance/exit, the `zone_id` is ignored. |  
+|  stop_url | URL | Optional | URL of a web page about a particular location. This should be different from the `agency_url` and the `route_url` field values. | 
+|  location_type | Enum | Optional | Indicates whether this `stop_id` represents a stop, station, or station entrance/exit. If `location_type` is omitted, all locations are treated as stops. Valid values are: |  
```
If you could us `<br>` instead of new cells, that would be amazing!

giocorti replied:
@LeoFrachet I'm a little confused. Do you want to use line breaks instead of cells to denote all enum options?

LeoFrachet replied:
It's really good! In the example you're linking above, you could also add the Example of direction_id in the same cell, that would be even better.

But now that I see it formated, I do think that the "ID from calendar.service_id" are taking to much place. @barbeau @antrim Thoughts?

giocorti replied:
I've added examples to the same cell using the <hr> tag as I think it gives them the proper visual weight. Heres a gist that shows the new enum option and example formatting. LMK what you all think.

[https://gist.github.com/giocorti/7eb24dab6c7a6546db44c23b02802d65](https://gist.github.com/giocorti/7eb24dab6c7a6546db44c23b02802d65)

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+|  location_type | Enum | Optional | Indicates whether this `stop_id` represents a stop, station, or station entrance/exit. If `location_type` is omitted, all locations are treated as stops. Valid values are: |  
+|   |  | | `0` or "empty" - Stop. A location where riders board or disembark from a transit vehicle. | 
+|   |  | | `1`- Station. A physical structure or area that contains one or more stop. |  
+|   |  | | `2` - Station Entrance/Exit. A point where riders can enter or exit a station from the street. The record must also specify a parent station by setting `parent_station` equal to the relevant `stop_id`. | 
```
Either put a space after the digit and before the dash, or do not. But here you do both (Yeah I'm that picky 😬).

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+|   |  | | `0` or "empty" - Stop. A location where riders board or disembark from a transit vehicle. | 
+|   |  | | `1`- Station. A physical structure or area that contains one or more stop. |  
+|   |  | | `2` - Station Entrance/Exit. A point where riders can enter or exit a station from the street. The record must also specify a parent station by setting `parent_station` equal to the relevant `stop_id`. | 
+|  parent_station | ID referencing `stops.stop_id` | Optional | Identifies the station that a stop or station entrance/exit is part of. This station-stop relationship, called a parent-child relationship, can only exist between a station (as the parent) and a stop or station entrance/exit (as the child). Recall that status as a station, stop or station entrance/exit is defined by `location_type`.  ...
```
> ID referencing `stops.stop_id`

I love it! Maybe we could use it everywhere!

LeoFrachet replied:
> a station entrance/exit must always have a parent.

Shall we say conditionnally required then?

barbeau replied:
`stops.stop_id` is a much better notation IMHO than saying "the stop_id field in the stops table", because the former is simply object and field notation and is format independent, where the latter is specific to CSV.  So yes, if Leo is suggest using this notation everywhere I agree.

LeoFrachet replied:
In fact it's already used in most places. But not in stops.zone_id for example.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+|   |  | | `1`- Station. A physical structure or area that contains one or more stop. |  
+|   |  | | `2` - Station Entrance/Exit. A point where riders can enter or exit a station from the street. The record must also specify a parent station by setting `parent_station` equal to the relevant `stop_id`. | 
+|  parent_station | ID referencing `stops.stop_id` | Optional | Identifies the station that a stop or station entrance/exit is part of. This station-stop relationship, called a parent-child relationship, can only exist between a station (as the parent) and a stop or station entrance/exit (as the child). Recall that status as a station, stop or station entrance/exit is defined by `location_type`.  ...
+|  stop_timezone | Timezone | Optional | Timezone of the location. If a location has a parent station, it inherits the parent station’s timezone instead of applying its own. Stations and parentless stops with `stop_timezone`="empty" inherit the timezone specified by `agency.agency_timezone`. If `stop_timezone` values are provided, the times in [stop_times.txt](#stop_timetxt) should be entered as  ...
```
Should be:

> empty `stop_timezone`

Because `stop_timezone`="empty" is really confusing IMHO.

LeoFrachet replied:
> If `stop_timezone` values are provided, the times in [stop_times.txt](#stop_timetxt) should be entered as the time since midnight

Is this new, or was it somewhere else in the spec? This is big. Why not noon-12 like everywhere else?

giocorti replied:
This isn't new. The current spec says
>Even if stop_timezone values are provided in stops.txt, the times in stop_times.txt should continue to be specified as time since midnight in the timezone specified by agency_timezone in agency.txt. This ensures that the time values in a trip always increase over the course of a trip, regardless of which timezones the trip crosses.

LeoFrachet replied:
Oh ok. My bad.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+|  parent_station | ID referencing `stops.stop_id` | Optional | Identifies the station that a stop or station entrance/exit is part of. This station-stop relationship, called a parent-child relationship, can only exist between a station (as the parent) and a stop or station entrance/exit (as the child). Recall that status as a station, stop or station entrance/exit is defined by `location_type`.  ...
+|  stop_timezone | Timezone | Optional | Timezone of the location. If a location has a parent station, it inherits the parent station’s timezone instead of applying its own. Stations and parentless stops with `stop_timezone`="empty" inherit the timezone specified by `agency.agency_timezone`. If `stop_timezone` values are provided, the times in [stop_times.txt](#stop_timetxt) should be entered as  ...
+|  wheelchair_boarding | Enum | Optional | Indicates whether wheelchair boardings are possible from the specified location. Valid values are: |
+|   |  | | For parentless stops: |
```
parentless location, no?

I would assume we are allowed to specify those values on entrances.

giocorti replied:
>parentless location, no?

I think it should stay as is. Entrances has its own separate enum options (although they are very similar to the parentless stop options). Also, I'm under the impression that entrances cannot be parentless, so parentless locations can only mean stations or parentless stops. Am I mistaken?
> |   |  | | For station entrances/exits: | 
|   |  | | `0` or "empty" - Station entrance/exit will inherit its `wheelchair_boarding` value from the parent station, if specified for the parent. |  
|   |  | | `1` - Station entrance/exit is wheelchair accessible. |  
|   |  | | `2` - No accessible path from station entrance/exit to stops/platforms. |

LeoFrachet replied:
Make sense.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+|   |  | | `0` or "empty" - No accessibility information for the stop. |
+|   |  | | `1` - Some vehicles at this stop can be boarded by a rider in a wheelchair. | 
+|   |  | | `2` - Wheelchair boarding is not possible at this stop. |  
+|   |  | | For child stops: |  
```
child stops? I would say for stops.

giocorti replied:
"Child stops" means any stop that is associated with a station. A child stop will inherit the `wheelchair_boarding` behavior from its parent station if `wheelchair_boarding`=`0` or empty. On the other hand, for a parentless stop (stop not in a station), `wheelchair_boarding`=`0` or empty means no wheelchair accessibility information is available. So I think we need to include both kinds of stops separately.

LeoFrachet replied:
Yes ok, my bad.

LeoFrachet commented:
@giocorti Amazing job, thanks! I only reviewed until end of `stops.txt`, but great job.

They are many small things to tighten up. But we're going somewhere.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
+|  friday | Enum | **Required** | Functions in the same way as `monday` except applies to Fridays  |
+|  saturday | Enum | **Required** | Functions in the same way as `monday` except applies to Saturdays. |
+|  sunday | Enum | **Required** | Functions in the same way as `monday` except applies to Sundays. |
+|  start_date | Date | **Required** | Start date for the service interval. |
```
If you add the definition of service day somewhere, you could improve this by saying: "Start *service day* for the service interval".

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
+|   |  | | `2` - Riders may transfer twice. |
+|   |  | | "empty" - Unlimited transfers are permitted. |
+|  agency_id | ID referencing `agency.agency_id` | **Conditionally Required** | Identifies the relevant agency for a fare. This field is required for datasets with multiple agencies defined in [agency.txt](#agencytxt), otherwise it is optional. |
+|  transfer_duration | Non-negative integer | Optional | Length of time in seconds before a transfer expires.  Should be "empty" or omitted when `transfers`=`0`, unless no transfers are allowed, in which case it can be used to indicates how long a ticket is valid for. |
```
I don't understand the "unless no transfers are allowed" here.

barbeau commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
-* **Field conditionally required** - This field or file is **required** under certain conditions, which are outlined in the field or file *Details*. Outside of these conditions, this field or file is optional. 
-* **Dataset unique** - The field contains a value that maps to a single distinct entity within the column. For example, if a route is assigned the ID **1A**, then no other route may use that route ID. However, you may assign the ID **1A** to a location because locations are a different type of entity than routes.
+* **Feed** - Successive datasets comprise a feed.
+* **Dataset** - A complete set of files defined by this specification reference. Datasets should be published at a public, permanent URL, including the zip file name. (e.g., www.agency.org/gtfs/gtfs.zip). For more information see GTFS best practices. 
```
Agreed, I'm not sure the "Feed" definition will be universally agreed-upon.  We should probably vote on that separately.

barbeau replied:
I'd expand the example to the full URL - `https://www.agency.org/gtfs/gtfs.zip`

antrim replied:
I could see how this would be debated, but I also am having a hard time seeing alternative definitions that are usable. The definitions we have proposed are based on existing usage, albeit quite inconsistent usage.

schema.org's uses of the terms "data feed" and "dataset" seem to match what we are doing:
- https://schema.org/DataFeed
- https://schema.org/Dataset

Basically, I think that we should be bold in this case. The terms "dataset" and "feed" have been consistently applied in this proposed spec modification. We could revert, but it hurts to leave this inconsistency in the Spec.

An idea: Prior to the pull request shall we bring the question to a somewhat broader group to see if they agree with our change?

barbeau commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,176 @@ This document explains the types of files that comprise a GTFS transit feed and
+* **Feed** - Successive datasets comprise a feed.
+* **Dataset** - A complete set of files defined by this specification reference. Datasets should be published at a public, permanent URL, including the zip file name. (e.g., www.agency.org/gtfs/gtfs.zip). For more information see GTFS best practices. 
+* **Field** - A specific type of information for records in a .txt file. Represented, in a table, as a column.
+* **Record** - A basic data structure comprised of a number of different fields describing a single entity (e.g. transit agency, stop, route, etc.). Represented, in a table, as a row. 
```
IMHO, a Record is a group of field values, not fields.

giocorti commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +211,183 @@ The example below is valid, with distinct blocks every day of the week.
+|  monday | Enum | **Required** | Indicates whether the service has been added or removed for all Mondays in the date range specified by the `start_date` and `end_date` fields. Note that exceptions for particular may be listed in [calendar_dates.txt](#calendar_datestxt). Valid values are: |
+|   |  | | `1` - Service is available for all Mondays in the date range. |
+|   |  | | `0` - Service is not available for Mondays in the date range. |
+|  tuesday | Enum | **Required** | Functions in the same way as `monday` except applies to Tuesdays |
```
I'm going to implent this suggestion unless someone has objections. I've added a screenshot if you're interested in seeing what it looks like. 

<img width="687" alt="screen shot 2018-11-05 at 3 25 53 pm" src="https://user-images.githubusercontent.com/31487717/48033346-df2fa200-e10f-11e8-9219-f04075405d06.png">

barbeau replied:
LGTM

LeoFrachet replied:
Looks good!

LeoFrachet commented:
I must confess that in all the proposals I'm writing, I put only two columns (Field name and description), and I just always start the description by the type and the cardinality:

It's more compact than with 4 columns. But I dunno. Some of you (IIRC) were in favor of the 4 columns.
@antrim @barbeau 

<img width="922" alt="capture d ecran le 2018-11-06 a 14 09 30" src="https://user-images.githubusercontent.com/3440724/48087578-b5d54b80-e1cd-11e8-938e-461cbef601b7.png">
<img width="686" alt="capture d ecran le 2018-11-06 a 14 09 25" src="https://user-images.githubusercontent.com/3440724/48087579-b5d54b80-e1cd-11e8-8361-1febf70bc780.png">

LeoFrachet commented:
@giocorti Could you also handle this request please Giovanni: https://github.com/google/transit/issues/90

barbeau commented:
@giocorti There is also a good discussion of frequencies.txt and frequency-based trips in https://github.com/google/transit/pull/47, including by @antrim and myself.  Could you please take a look at that and see if it makes sense to pull any of that proposed content into this PR?

barbeau commented:
> Some of you (IIRC) were in favor of the 4 columns.

IIRC GitHub's Markdown renderer shows columns with equal width no matter the content, so 2 columns ends up wasting a lot of space when viewing on GitHub.  Also, generally, I think 4 columns is easier to digest for newcomers to the spec, so I'd suggest keeping that.

Idea - in the HTML-rendered site, we could offer a "collapsed" view for pros that shows only two columns in the format the Leo is using.  I think that would be relatively straightforward (?) if we encode the info in 4 columns in Markdown.

giocorti commented:
I've reviewed all the suggestions and made the changes summarized below. Thanks for all the input! LMK what you think. 

* Minor editorial changes
* Empty fields are now considered to have no value 
* Rewrote “Optional" definition
* Rewrote “field” definition
* Removed "field" (This may be changed after more discussion)
* URLs not hidden
* Defined “Service Day” 
* Reformatted enum and examples using `<br>` and `<hr>`
* Added code ticks to “Field Name” column
* Rewrote `transfer_duration`
* Moved google fares example to an internal .md file
* Rewrote frequencies.txt and `exact_times` descriptions

LeoFrachet commented:
I'm rereading it completely before pushing it to the public repo.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -29,174 +29,144 @@ This document explains the types of files that comprise a GTFS transit feed and
-|   |  | | * **0** (or empty) - the station entrance will inherit its **wheelchair_boarding** value from the parent station, if specified in the parent |  
-|   |  | | * **1** - the station entrance is wheelchair accessible (e.g. an elevator is available to platforms if they are not at-grade)  |  
-|   |  | | * **2** - there exists no accessible path from the entrance to station platforms |  
+|  `stop_id` | ID | **Required** | Identifies a stop, station, or station entrance. Stops, stations or station entrances/exits are collectively referred to as locations. Multiple routes may use the same stop. | 
```
> Identifies a stop, station or station entrance/exit.

If you wanna be consistent.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +179,152 @@ The example below is valid, with distinct blocks every day of the week.
+|  `stop_id` | ID referencing `stops.stop_id` | **Required** | Identifies the serviced stop. All stops serviced during a trip must have a record in [stop_times.txt](#stop_timestxt). Referenced locations must be stops, not stations or station entrances/exits. A stop may be serviced multiple times in the same trip, and multiple trips and routes may service the same stop. |  
+|  `stop_sequence` | Non-negative integer | **Required** | Order of stops for a particular trip. The values must increase along the trip but do not need to be consecutive.<hr>*Example: The first location on the trip could have a `stop_sequence`=`1`, the second location on the trip could have a `stop_sequence`=`23`, the third location could have a `stop_sequence`=`40`, and so on.* |  
+|  `stop_headsign` | Text | Optional | Text that appears on signage identifying the trip's destination to riders. This field overrides the default `trips.trip_headsign` when the headsign changes between stops. If the headsign is displayed for an entire trip, use `trips.trip_headsign` instead. |  
+|  `pickup_type` | Enum | Optional | Indicates drop off method. Valid options are:<br><br>`0` or "empty - Regularly scheduled pickup. <br>`1` - No pickup available.<br>`2` - Must phone agency to arrange pickup.<br>`3` - Must coordinate with driver to arrange pickup. |  
```
"empty => double quote should be removed on the left side.

LeoFrachet commented on gtfs/spec/en/reference.md:
```diff
@@ -209,244 +179,152 @@ The example below is valid, with distinct blocks every day of the week.
+|  `stop_sequence` | Non-negative integer | **Required** | Order of stops for a particular trip. The values must increase along the trip but do not need to be consecutive.<hr>*Example: The first location on the trip could have a `stop_sequence`=`1`, the second location on the trip could have a `stop_sequence`=`23`, the third location could have a `stop_sequence`=`40`, and so on.* |  
+|  `stop_headsign` | Text | Optional | Text that appears on signage identifying the trip's destination to riders. This field overrides the default `trips.trip_headsign` when the headsign changes between stops. If the headsign is displayed for an entire trip, use `trips.trip_headsign` instead. |  
+|  `pickup_type` | Enum | Optional | Indicates drop off method. Valid options are:<br><br>`0` or "empty - Regularly scheduled pickup. <br>`1` - No pickup available.<br>`2` - Must phone agency to arrange pickup.<br>`3` - Must coordinate with driver to arrange pickup. |  
+|  `drop_off_type` | Enum | Optional | Indicates pick up method. Valid options are:<br><br>`0` or "empty - Regularly scheduled drop off.<br>`1` - No drop off available.<br>`2` - Must phone agency to arrange drop off.<br>`3` - Must coordinate with driver to arrange drop off. |    
```
> "empty

=> double quote should be removed on the left side.

LeoFrachet commented:
For shape, the text is cut because of the quote. Would there be a way to solve it?

<img width="955" alt="capture d ecran le 2018-11-16 a 16 09 42" src="https://user-images.githubusercontent.com/3440724/48647484-17f63380-e9ba-11e8-9778-d1fa1a31eaa1.png">

LeoFrachet commented:
LGTM, with the three small comments above.

giocorti commented:
> For shape, the text is cut because of the quote. Would there be a way to solve it?
> 
> <img alt="capture d ecran le 2018-11-16 a 16 10 30" width="1788" src="https://user-images.githubusercontent.com/3440724/48647503-2b090380-e9ba-11e8-87ad-f4f36a84f669.png">

My browser allows side to side scrolling on the shapes.txt table. Certainly not ideal but, as far as I can tell, its best we can do if we want to keep the last example.
